Move autosubmit-config-parser back into autosubmit

kinow opened the pull request:
Partially addresses #2052 

Changes done with IDE.

First I updated the pyproject.toml of Autosubmit, copying what was in autosubmit-config-parser.

Then I reviewed if the EasyBuild recipe needed to be updated (I didn't update ours, as the only difference to the autosubmit-config-parser was that we include ruamel in `exts_list`, and there we had ruamel in `dependencies`. I believe in the end both are installed and available as Python libs?

After that, it was just a matter of moving the files from `autosubmitconfig.config` to `autosubmit.config`, and using "Replace in Files" in PyCharm to replace everything. The only occurrence of `autosubmitconfigparser` that I did not remove was one in the changelog, as it's referencing the autosubmit-config-parser GitHub repository (which will be archived, not deleted).

I moved all unit tests, but some were not running (referencing $HOME directories from a user, others broken due to changes). I fixed the tests I could, but had to skip the ones that references user data (@dbeltrankyl I think if we include the workflows in the repo we can fix those tests, but maybe those can be fixed with `autosubmit_exp` creating simple workflows).

~The regression folder had a destine workflow example workflow, but it didn't appear to be running anything. I left it in autosubmit-config-parser, but we can move it here later if needed.~ I had missed the Python scripts, thinking there were only YAML files in that folder. Thanks for spotting that, @dbeltrankyl. I've added the regression YAML and Python scripts back here.

@dbeltrankyl we might have to update easybuild/module installations. It this is merged, we need to archive the autosubmit-config-parser repository, and then I will move all issues to this repository, and see which pull requests can be moved too (I would just push to a branch here, adjusting the imports, and open a pull request).

**Check List**

- [x] I have read `CONTRIBUTING.md`.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to `pyproject.toml`.
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included in `CHANGELOG.md` if this is a change that can affect users.
- [x] Documentation updated.
- [x] If this is a bug fix, PR should include a link to the issue (e.g. `Closes #1234`).

kinow commented:
Interesting. First failure on GH actions was only for unit tests (regression/integration appeared to be OK). And it was because we had a dependency on `autosubmit-config-parser`, which had a dependency on `mock`.

Some of our unit tests depended on `mock` (not our main code), so I just had to add that to the `tests` optional group of dependencies, with a note that that's not necessary as we can use `pytest-mock` instead. The build should pass now, and coverage will go down. I will take note of how much it decreased, and then add the tests of autosubmit-config-parser, and wait for Codecov to report again.

kinow commented:
Very happy to be wrong!!! @isimo00 @dbeltrankyl @VindeeR , moving the code + tests to this repository resulted in a slight increase in our test coverage :)

<img width="970" height="1065" alt="image" src="https://github.com/user-attachments/assets/9e4679a6-9706-498f-b532-ec840a7e0391" />

(Although I found some parts of the config-parser code that appear to have bugs, one I saved here to open an issue, but feel free to have a look at the code if you'd like.)

We still have to improve coverage, but good that it didn't reduce our coverage total.

dbeltrankyl commented:
Hello @kinow 

> The regression folder had a destine workflow example workflow

The regression/test_reload.py script uses these files to function (I just tried running them, and it works). You didn't move it, no? Or at least I didn't see it

kinow commented:
> Hello @kinow
> 
> > The regression folder had a destine workflow example workflow
> 
> The regression/test_reload.py script uses these files to function (I just tried running them, and it works). You didn't move it, no? Or at least I didn't see it
> <img alt="image" width="412" height="199" src="https://private-user-images.githubusercontent.com/94113100/470793199-83a84572-7ab0-423a-a90d-bef8555b8ba4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM0NDg3OTcsIm5iZiI6MTc1MzQ0ODQ5NywicGF0aCI6Ii85NDExMzEwMC80NzA3OTMxOTktODNhODQ1NzItN2FiMC00MjNhLWE5MGQtYmVmODU1NWI4YmE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI1VDEzMDEzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAwYWExZTk0YzQxNzdkYWJkNjkwNTAyZmU2ZTk1ZWRjNmU3ZDhjNmI0MjIxYTAxMjYxMTkwOWI3MjdmMzZjYzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.H02mI3exMxBzB0H2rIITpQ1seVdH5wPbxzVobJ08uZ8">

Ah, I missed that one. Give me some minutes and I will add it back...

kinow commented:
Added a new commit doing this:

- [x] added license headers
- [x] formatted with PyCharm
- [x] optimized imports (sorting)
- [x] fixed grammar/typos
- [x] fixed mutable default argument (when `def fn(a_list: list())`, where `a_list` can retain state among calls)
- [x] added some mypy type hints
- [x] fixed other minor issues like pydoc, style, etc.

Left as a separate commit in case we need to revert it.

kinow commented:
Done, @dbeltrankyl ! Regression test added back.

dbeltrankyl reviewed (approved):
Thanks @kinow for doing this!


For me it is ok to merge it

kinow commented:
> Thanks @kinow for doing this!
> 
> For me it is ok to merge it

Then let's end the week with the code merged! Merging now, thanks @dbeltrankyl !!!

kinow commented:
Commits squashed. Waiting for GH Actions...
